Plutarch/Builtin.hs

```diff
@@ -1,27 +1,122 @@
-module Plutarch.Builtin (PData (..), pfstBuiltin, psndBuiltin, pasConstr, PBuiltinPair, PBuiltinList) where
+module Plutarch.Builtin (
+  PData (..),
+  pTrace,
+  (#£),
+  (!£),
+  PBuiltin (..),
+  PList (..),
+  PPair (..),
+) where
 
-import Plutarch (punsafeBuiltin)
+import Data.Proxy
+import Data.Text (Text)
+import Data.Type.Nat
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin.List.Type
+import Plutarch.Builtin.Pair.Type
 import Plutarch.ByteString (PByteString)
 import Plutarch.Integer (PInteger)
 import Plutarch.Prelude
+import Plutarch.String
 import qualified PlutusCore as PLC
 
-data PBuiltinPair (a :: k -> Type) (b :: k -> Type) (s :: k)
-
-data PBuiltinList (a :: k -> Type) (s :: k)
-
 data PData s
-  = PDataConstr (Term s (PBuiltinPair PInteger (PBuiltinList PData)))
-  | PDataMap (Term s (PBuiltinList (PBuiltinPair PData PData)))
-  | PDataList (Term s (PBuiltinList PData))
+  = PDataConstr (Term s (PPair PInteger (PList PData)))
+  | PDataMap (Term s (PList (PPair PData PData)))
+  | PDataList (Term s (PList PData))
   | PDataInteger (Term s PInteger)
   | PDataByteString (Term s PByteString)
 
-pfstBuiltin :: Term s (PBuiltinPair a b :--> a)
-pfstBuiltin = phoistAcyclic $ pforce . pforce . punsafeBuiltin $ PLC.FstPair
+{- | Type spec for PLC's untyped builtin functions
+
+ Note: `forces` determines the repeated application of `FORCE` when evaluating
+ a polymorphic builtin function.
+
+ Example: (UnConstrData #£ someData)
+-}
+data PBuiltin (forces :: Nat) (args :: [k -> Type]) (res :: k -> Type) where
```

Comment:
Why do we need this?

Reply:
Type-safety. Automatically gives us typed builtin functions. For example,

```haskell
chooseData ::
  forall k (s :: k) (c' :: k -> Type).
  Term s PData ->
  Term s (PDelayed c' :--> PDelayed c' :--> PDelayed c' :--> PDelayed c' :--> PDelayed c' :--> PDelayed c')
chooseData = (B.ChooseData #£)
```

Compare to untyped:

```haskell
chooseDataUntyped :: forall k (s :: k) (a :: k -> Type). Term s a
chooseDataUntyped = punsafeBuiltin PLC.ChooseData
```

Plus, the GADT also encodes the number of forces required at the type-level.

Reply:
I don't see how this is better than writing the typed functions manually.

Reply:
Even if you explicitly add the type of `Term` to `chooseDataUntyped` and other builtins -- and I doubt people will do that (given the 21 instances of untyped uses at the moment[^1]) --  callers would still have to remember to `force` the builtin.

I find it much more ergonomic and safer to use this.

[^1]: .
       ```
       ❯ rg -I 'punsafeBuiltin PLC\.' | wc -l
       21
       ```

Reply:
I'm not sure what you mean? This is the approach we've been using until now: https://github.com/Plutonomicon/plutarch/pull/20/files#diff-74f4e5889ff8a668263439d5d956f02e7f56dcf0cf7f27987f3e2e4d0672d60eR32

Reply:
No, that's from the PR (`main` branch still uses untyped ones). 

I think @TotallyNotChase's approach begins to address the same problem, but it is a matter of preference as to whether you want to write `(B.TailList #£)` or `ptailBuiltin`. I personally prefer the former, because I don't have to learn yet another literal, and can just reuse the Plutus Core's naming. Regardless, `punsafeBuiltin` should not be exposed outside of where the typed builtins are defined.

Reply:
https://github.com/Plutonomicon/plutarch/blob/9d08fcfc4b9cf260fd197cb7688a0a29b43d71cc/Plutarch/Bool.hs#L31

Reply:
Sure - and there are 21 instances of untyped. Why? :-P My point still stands.